### PR TITLE
CUMULUS-2306: Update executions 'get' endpoint to use postgres database

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-XX: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-XXX)

## Changes

* Detailed list or prose of changes
* ...

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

### CODE REVIEW

File: packages/api/bin/serve.js [hunk 1]
```diff
@@ -165,28 +184,41 @@ async function initializeLocalElasticsearch(stackName) {
 }
 
 /**
- * Fill dynamo and elastic with fake records for testing.
+ * Fill dynamo, postgres and elastic with fake records for testing.
  * @param {string} stackName - The name of local stack. Used to prefix stack resources.
  * @param {string} user - username
  */
-async function createDBRecords(stackName, user) {
+async function createDBRecords(stackName, user, knex) {
```
Comment (Jkovarik) [inline packages/api/bin/serve.js:191]: We're going to need to update this on each ticket, in theory.

File: packages/api/endpoints/execution-status.js [hunk 2]
```diff
@@ -82,9 +86,10 @@ async function get(req, res) {
...
   let response;
-  const e = new models.Execution();
+  const knex = await getKnexClient({ env: process.env });
+  const executionPgModel = new ExecutionPgModel();
   try {
-    response = await e.get({ arn });
+    response = await executionPgModel.get(knex, { arn });
   } catch (error) {
     if (error instanceof RecordDoesNotExist) {
       return res.boom.notFound('Execution not found in API or database');
```
Comment (npauzenga) [inline packages/api/endpoints/execution-status.js:95]: Would it be clearer to return the error that `get` throws? 
`Record in ${this.tableName} with identifiers ${JSON.stringify(params)} does not exist.`

File: packages/api/endpoints/executions.js [hunk 3]
```diff
@@ -31,18 +36,20 @@ async function list(req, res) {
...
+  const knex = await getKnexClient({ env: process.env });
+  const executionPgModel = new ExecutionPgModel();
+  let executionRecord;
   try {
-    const response = await e.get({ arn });
-    return res.send(response);
+    executionRecord = await executionPgModel.get(knex, { arn });
   } catch (error) {
     if (error instanceof RecordDoesNotExist) {
       return res.boom.notFound(`No record found for ${arn}`);
```
Comment (npauzenga) [inline packages/api/endpoints/executions.js:46]: Same as above, we could just return the message from `RecordDoesNotExist`. Super minor though. Either makes sense.

File: packages/api/tests/endpoints/test-executions.js [hunk 4]
```diff
@@ -57,6 +71,7 @@ test.before(async () => {
   // create fake execution records
   fakeExecutions.push(fakeExecutionFactory('completed'));
   fakeExecutions.push(fakeExecutionFactory('failed', 'workflow2'));
+  // TODO - this needs updated after postgres->ES work
```
Comment (npauzenga) [inline packages/api/tests/endpoints/test-executions.js:74]: Oh because the source of the ES index needs to be PG. Do we have that work planned?

File: packages/api/tests/endpoints/test-executions.js [hunk 5]
```diff
@@ -147,21 +179,86 @@ test('executions can be filtered by workflow', async (t) => {
...
+  await executionPgModel.create(
+    t.context.knex,
+    executionRecord
+  );
+  t.teardown(async () => {
+    await executionPgModel.delete(t.context.knex, executionRecord);
+    await executionPgModel.delete(t.context.knex, parentExecutionRecord);
+    await collectionPgModel.delete(t.context.knex, collectionRecord);
+    await asyncOperationsPgModel.delete(t.context.knex, asyncRecord);
+  });
```
Comment (npauzenga) [inline packages/api/tests/endpoints/test-executions.js:220]: Does teardown work in reverse order?

File: packages/api/tests/endpoints/test-executions.js [hunk 6]
```diff
@@ -147,21 +179,86 @@ test('executions can be filtered by workflow', async (t) => {
...
+
+  const executionResult = response.body;
+  const expectedRecord = await translatePostgresExecutionToApiExecution(
+    executionRecord,
+    t.context.knex
+  );
+
+  t.is(executionResult.arn, executionRecord.arn);
+  t.is(executionResult.asyncOperationId, asyncRecord.id);
+  t.is(executionResult.collectionId, `${collectionRecord.name}___${collectionRecord.version}`);
```
Comment (npauzenga) [inline packages/api/tests/endpoints/test-executions.js:236]: Same comment as above. We have a helper for this if it makes sense to import here. `constructCollectionId`

File: packages/api/tests/endpoints/test-executions.js [hunk 7]
```diff
@@ -147,21 +179,86 @@ test('executions can be filtered by workflow', async (t) => {
...
+  t.is(executionResult.parentArn, parentExecutionRecord.arn);
+  t.like(executionResult, expectedRecord);
+});
+
+test('GET returns an existing execution without any foreign keys', async (t) => {
```
Comment (npauzenga) [inline packages/api/tests/endpoints/test-executions.js:241]: Do we need this test? 🤔 

It seems like `get` will pull whatever is in the DB and not really care about foreign keys. When we create or delete an execution we definitely want to check with and without but maybe not on `get`?

File: packages/db/src/translate/executions.ts [hunk 8]
```diff
@@ -3,10 +3,68 @@ import Knex from 'knex';
...
+): Promise<ExecutionRecord> => {
+  let parentArn: string | undefined;
+  let collectionId: string | undefined;
+  let asyncOperationId: string | undefined;
+
+  if (executionRecord.collection_cumulus_id) {
+    const collection = await collectionPgModel.get(knex, {
+      cumulus_id: executionRecord.collection_cumulus_id,
+    });
+    collectionId = `${collection.name}___${collection.version}`;
```
Comment (npauzenga) [inline packages/db/src/translate/executions.ts:27]: Same comments as above. We could use a helper here if it's not a pain to import. `constructCollectionId`

File: packages/db/src/translate/executions.ts [hunk 9]
```diff
@@ -3,10 +3,68 @@ import Knex from 'knex';
...
+    asyncOperationId = asyncOperation.id;
+  }
+  if (executionRecord.parent_cumulus_id) {
+    const parentExecution = await executionPgModel.get(knex, {
+      cumulus_id: executionRecord.parent_cumulus_id,
+    });
+    parentArn = parentExecution.arn;
+  }
+
+  const name = executionRecord.arn.split(':').pop();
```
Comment (npauzenga) [inline packages/db/src/translate/executions.ts:42]: Nit: Could we comment about what the intended result looks like or make this a helper method with an example?

Comment (npauzenga) [reply packages/db/src/translate/executions.ts:42]: Oh or maybe just note that it's the `postfix`. That might be clear enough.

File: packages/db/tests/translate/test-executions.js [hunk 10]
```diff
@@ -2,7 +2,79 @@ const test = require('ava');
 const cryptoRandomString = require('crypto-random-string');
 
 const { removeNilProperties } = require('@cumulus/common/util');
-const { translateApiExecutionToPostgresExecution } = require('../../dist/translate/executions');
+const {
+  translateApiExecutionToPostgresExecution,
+  translatePostgresExecutionToApiExecution,
+} = require('../../dist/translate/executions');
+
+test.only('translatePostgresExecutionToApiExecution translates a Postgres execution to an API record', async (t) => {
```
Comment (npauzenga) [inline packages/db/tests/translate/test-executions.js:10]: Reminder to remove `only` 😄

**Review by npauzenga — CHANGES_REQUESTED** [review]
Looks great @Jkovarik! Just a few comments/questions.

File: packages/api/bin/serve.js [hunk 11]
```diff
@@ -165,45 +183,79 @@ async function initializeLocalElasticsearch(stackName) {
...
   const execution = testUtils.fakeExecutionFactoryV2({ arn: `${stackName}-fake-arn` });
   await serveUtils.addExecutions([execution]);
 
+  const postgresExecution = await translateApiExecutionToPostgresExecution(execution, knex);
+  await executionPgModel.upsert(knex, postgresExecution);
+
+  // add fake granule records
+  const granule = testUtils.fakeGranuleFactoryV2({
+    granuleId: `${stackName}-granule`,
+    collectionId: `${collection.name}___${collection.version}`,
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/bin/serve.js:251]: 🤦🏻  I always shorthand this and then intend to call the helper and forget.   Good catch.

File: packages/api/tests/endpoints/test-executions.js (hunk 6 above)
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-executions.js:236]: Agreed - it's always a pain to lookup, so I shorthand with the intent to go back and sort, and forgot it here.      Good catch.

File: packages/api/bin/serve.js [hunk 12]
```diff
@@ -165,45 +183,79 @@ async function initializeLocalElasticsearch(stackName) {
...
+    name: `${stackName}-collection`,
+    version: '0.0.0',
+  });
   await serveUtils.addCollections([collection]);
-
-  // add granule records
-  const granule = testUtils.fakeGranuleFactoryV2({ granuleId: `${stackName}-granule` });
-  await serveUtils.addGranules([granule]);
+  const postgresCollection = await translateApiCollectionToPostgresCollection(collection);
+  await collectionPgModel.upsert(knex, postgresCollection);
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/bin/serve.js:215]: Possibly. 

   I was opting for a minimum changeset here, my goal mostly was to ensure that what was already being populated was done within a ticket-scope-limited way, given it's not entirely clear to me who's using this outside of local-api test runs to debug/etc the dashboard.       Strictly, the scope of this was to make sure the executions endpoint run from this set of commands still shows an execution.
   
tl;dr yes, but I think that's probably legitimately out of scope.

[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/bin/serve.js:215]: To some degree it depends on what this is being used for.     If it's called as part of serve, the tables may not be wiped.


   We could improve what it's doing to delete/add, but that's a change in the current behavior/from what I believe the dynamo methods are doing (the `.create` call in the model eventually calls DynamoDbDocClient.put, which `* Creates a new item, or replaces an old item with a new item by delegating to AWS.DynamoDB.putItem().`, and probably exceeds the scope of the ticket.      That is, of course, up for debate.

File: packages/api/bin/serve.js [hunk 13]
```diff
@@ -165,45 +183,79 @@ async function initializeLocalElasticsearch(stackName) {
...
+    provider: provider.name,
     collection: {
-      name: `${stackName}-collection`,
-      version: '0.0.0',
+      name: collection.name,
+      version: collection.version,
+    },
+    rule: {
+      type: 'onetime',
+      arn: 'someArn',
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/bin/serve.js]: 👍🏻 a2476aaff

File: packages/db/tests/translate/test-executions.js (hunk 10 above)
Comment (Jkovarik) [reply packages/db/tests/translate/test-executions.js:10]: Ugh.  I had to remove the commit hook because we *have* tests that deliberately *are* .only :|

File: packages/api/tests/endpoints/test-executions.js (hunk 6 above)
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-executions.js]: f13521635

File: packages/api/bin/serve.js [hunk 14]
```diff
@@ -314,10 +366,42 @@ async function serveDistributionApi(stackName = localStackName, done) {
...
+    providerPgModel.delete(knex, {}),
+  ];
+  await Promise.all(delPromises);
+}
+
 /**
  * erase all dynamoDB tables
  * @param {string} stackName - stack name (generally 'localrun')
- * @param {string} systemBucket - stystem bucket (generally 'localbucket' )
+ * @param {string} systemBucket - system bucket (generally 'localbucket'
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/bin/serve.js]: f13521635

File: packages/api/bin/serve.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/bin/serve.js]: f13521635

File: packages/api/endpoints/executions.js (hunk 3 above)
Comment (Jkovarik) [reply packages/api/endpoints/executions.js]: e7ff2de35

File: packages/api/endpoints/execution-status.js (hunk 2 above)
Comment (Jkovarik) [reply packages/api/endpoints/execution-status.js]: I'm assuming this means the text in the res.boom.notFound call, not returning the error itself: e7ff2de35

File: packages/api/tests/endpoints/test-executions.js (hunk 4 above)
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-executions.js:75]: Yes, there's a 'index postgres writes to ES' ticket, if that's what you mean.

File: packages/api/tests/endpoints/test-executions.js (hunk 5 above)
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-executions.js:221]: Interestingly enough, at some point tear down defaults to reverse order, but iirc that's *per block* meaning this shouldn't *need* to be?

https://github.com/nasa/cumulus/pull/2263/files#r646627149

Comment (Jkovarik) [reply packages/api/tests/endpoints/test-executions.js:221]: I'm not sure I'm following though - *should* it work in reverse order?

File: packages/api/tests/endpoints/test-executions.js (hunk 7 above)
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-executions.js:245]: It's fair that this isn't *really* behavior on the execution *endpoint* directly, this test was written for a failure  that eventually got abstracted to logic in `translatePostgresExecutionToApiExecution`.      

If translate isn't written to not make the subquery, or with a proper join (we shouldn't do this anyway), you get either no result (bad join logic) or a failure (what happens if you remove the current branch logic in the translate method)

File: packages/db/src/translate/executions.ts (hunk 8 above)
Comment (Jkovarik) [reply packages/db/src/translate/executions.ts]: f135216

File: packages/db/src/translate/executions.ts (hunk 9 above)
Comment (Jkovarik) [reply packages/db/src/translate/executions.ts:43]: It's probably not unreasonable to make this a helper, we're doing it *all over the place*.   I erred on the side of not adding it to @common given the idiom itself (if not the intent) is straightforward, but that's perhaps a mistake. 

In either case, this probably helps: edd75a016

File: packages/db/tests/translate/test-executions.js (hunk 10 above)
Comment (Jkovarik) [reply packages/db/tests/translate/test-executions.js]: edd75a016

File: packages/api/bin/serve.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/bin/serve.js:218]: Yup, totally agree.

[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/bin/serve.js:218]: Ah ok. That does appear to be the existing logic from Dynamo so agreed that this is the right approach here. I don't think it totally matters either way as (I don't think) we have problems with the test data being populated when serving the API locally.

File: packages/api/bin/serve.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (npauzenga) [reply packages/api/bin/serve.js]: )

File: packages/api/endpoints/execution-status.js (hunk 2 above)
Comment (npauzenga) [reply packages/api/endpoints/execution-status.js]: Ah yeah, I did mean the message. Could you get it from `error.message` though rather than adding the same text? The idea was just that we could avoid replacing one perfectly decent error message with another. 

```
if (error instanceof RecordDoesNotExist) {      
  return res.boom.notFound(error.message);
}
```

File: packages/api/endpoints/executions.js (hunk 3 above)
Comment (npauzenga) [reply packages/api/endpoints/executions.js]:
```
if (error instanceof RecordDoesNotExist) {      
  return res.boom.notFound(`Execution ${error.message}`);
}
```
But I don't feel super strongly about it 😄

File: packages/api/tests/endpoints/test-executions.js (hunk 4 above)
Comment (npauzenga) [reply packages/api/tests/endpoints/test-executions.js:75]: Awesome. Yup! That's what I was after.

File: packages/api/tests/endpoints/test-executions.js (hunk 5 above)
Comment (npauzenga) [reply packages/api/tests/endpoints/test-executions.js:221]: Huh. I must have read this wrong. What you have here is correct where executions are deleted before collections. For some reason I thought it was reversed.

File: packages/api/tests/endpoints/test-executions.js (hunk 7 above)
Comment (npauzenga) [reply packages/api/tests/endpoints/test-executions.js:245]: Yeah, thanks for doing that! Totally makes sense.

**Comment by npauzenga** [comment]
Thanks @Jkovarik! This looks great! I added one comment about possibly using the model's error message directly in the API response but that's a pretty minor nit. Take it or leave it 😄 .

**Review by npauzenga — APPROVED** [review]
